### ISSUE TITLE
Metadata request does not take into account query string

### ISSUE THREAD

**Issue by rouacke** [object/Simple.OData.Client#128]
SAP NetWeaver Gateway provides OData Service endpoints that take the following form:

https://example.com/sap/opu/odata/IWBEP/GWSAMPLE_BASIC/?sap-client=100

Note the query string. In SAP, "client" is the top-most organizational structure. Everything is effectively keyed by client--including, apparently, OData Service endpoints. The default client is `100`. In my experience, few installs have more than one client.

If I construct an instance of `ODataClientSettings` with the aforementioned URL, create an `ODataClient` and execute `GetMetadataDocumentAsync`, I see (using Fiddler) an HTTP request to the following URL:

https://example.com/sap/opu/odata/IWBEP/GWSAMPLE_BASIC/?sap-client=100/$metadata

Note that `$metadata` has been appended to the query string rather than the path. It appears that Simple.OData.Client appends `$metadata` without taking into account the query string. As a result, the client gets back an `app:service` XML document, not an `<edmx:Edmx>` document. At which point, Simple.OData.Client throws the following exception:

> The metadata document could not be read from the message content.
 UnexpectedXmlElement : The element 'app:service' was unexpected for the root element. The root element should be Edmx. : (0, 0)

I've confirmed that the following URL returns the metadata correctly:

https://example.com/sap/opu/odata/IWBEP/GWSAMPLE_BASIC/$metadata?sap-client=100

Here's a sample HTTP request:

```
GET https://example.com/sap/opu/odata/IWBEP/GWSAMPLE_BASIC/?sap-client=100/$metadata HTTP/1.1
x-csrf-token: Fetch
Authorization: Bearer {OAuth Bearer Token}
Accept: application/atom+xml, application/xml, application/text
Host: example.com
Connection: Keep-Alive
```

And the HTTP response:

```
HTTP/1.1 200 OK
set-cookie: sap-usercontext=sap-client=100; path=/
set-cookie: sap-XSRF_SE1_100={CSRF Token}; path=/; HttpOnly
content-type: application/xml
content-length: 4498
x-csrf-token: {CSRF Token}
sap-metadata-last-modified: Thu, 16 Jul 2015 21:54:37 GMT
cache-control: no-store, no-cache, must-revalidate, max-age=0, post-check=0, pre-check=0
pragma: no-cache
expires: Tue, 03 Jul 2001 06:00:00 GMT
last-modified: Fri, 31 Jul 2015 18:47:28 GMT
dataserviceversion: 2.0

<?xml version="1.0" encoding="utf-8"?><app:service xml:lang="en" xml:base="https://example.com/sap/opu/odata/IWBEP/GWSAMPLE_BASIC/" xmlns:app="http://www.w3.org/2007/app" xmlns:atom="http://www.w3.org/2005/Atom" xmlns:m="http://schemas.microsoft.com/ado/2007/08/dataservices/metadata" xmlns:sap="http://www.sap.com/Protocols/SAPData">{...snip...}</app:service>
```

As a workaround, I can simply remove the query string. The `sap-client` parameter does not seem to be required. I assume this is because the SAP NetWeaver Gateway system I happen to be connecting to has only one SAP client.

**Comment by object** [comment]
Thank you for reporting the issue. How are regular OData requests built, especially those that use query string? For example, if you query resource Products with a filter:

https://example.com/sap/opu/odata/IWBEP/GWSAMPLE_BASIC/Products?$filter=Name eq 'abc'

Is sap-client=100 expected to be attached using "&" operator? E.g.:

https://example.com/sap/opu/odata/IWBEP/GWSAMPLE_BASIC/Products?$filter=Name eq 'abc'&sap-client=100

It's pretty straightforward for me to make such fix, I would like to confirm it first.

**Comment by rouacke** [comment]
Thanks for considering this issue, Vagif. I've confirmed that the following request does work:

https://example.com/sap/opu/odata/IWBEP/GWSAMPLE_BASIC/ProductSet?$filter=TypeCode+eq+%27AD%27&sap-client=100

Note that I had to modify the URL slightly from your example. I queried the collection `ProductSet`, not the entity type `Product`. Also, the `Name` column is not filterable, so I used `TypeCode`. Let me know if this test isn't valid.

I don't have a access to a multi-client NetWeaver environment. So, I can't fully test this. However, if I start with the URL above and change `sap-client` to `200` (an invalid value), I get an authentication challenge:

> SAP NetWeaver Application Server [SE1/200]

Note the "200". That corresponds to the `sap-client` parameter. So, it appears NetWeaver honors the value of the `sap-client` parameter.

As a side note, I've noticed that NetWeaver sets a cookie named `sap-usercontext` with a value of `sap-client=100`. As a workaround, we might be able to strip the URL parameter from the OData service endpoint and append the cookie.

**Comment by object** [comment]
Thank you for the information. Yes, using cookie or custom HTTP header can be alternatives in such scenarios. But since the library behavior is erratic (it should not construct multiple query strings), I will make a change that will hopefully fix your issue. I'll let you know when you can test it.

**Comment by object** [comment]
I made a fix, and there is a pre-release of the version 4.9, available at MyGet:  https://www.myget.org/F/simple-odata-client/Packages 

It should support OData service baseUri containing query string. Can you check it out? I believe it should fix your issue.

**Comment by rouacke** [comment]
I've confirmed that the fix works. I tested the metadata query as well as CRUD operations. In all cases, Simple.OData.Client preserved the `sap-client` parameter.

Thank you very much.

**Comment by object** [comment]
Great! Thanks for testing it. Version 4.9 is on its way.

**Comment by SiriMB** [comment]
Is this issue of $metadata being appended to the odata URL with query parameters resolved ? I installed version 4.9 as well as the latest (4.14.2) and still see the issue.
Can you please help ?
Thanks.

**Comment by object** [comment]
Yes it was resolved (and confirmed) few versions ago. In case you are
experiencing a problem I need more details.

2016-02-05 19:44 GMT+01:00 SiriMB <notifications@github.com>:
